File: algorithms/86. Partition List/solution.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode left(0), right(0);
        ListNode * l = & left, * r = & right;
        while ( head ) {
            auto & p = head->val < x ? l : r;
            p = p->next = head;
            head = head->next;
        }
        r->next = NULL;
        l->next = right.next;
        return left.next;
    }
};

File: algorithms/86. Partition List/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        lt = lttail = ListNode(0)
        ge = getail = ListNode(0)
        while head:
            if head.val < x:
                lttail.next = lttail = head
                #lttail = head
            else:
                getail.next = getail = head
                #getail = head
            head = head.next
        lttail.next = ge.next
        getail.next = None
        return lt.next
